src/telemetry.ts

```diff
@@ -1,137 +1,193 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
 import * as vscode from 'vscode';
 import TelemetryReporter from 'vscode-extension-telemetry';
 
+import {CancelOperationError} from './CancelOperationError';
 import {DevelopEnvironment} from './constants';
 import {ExceptionHelper} from './exceptionHelper';
 import {RemoteExtension} from './Models/RemoteExtension';
 import {NSAT} from './nsat';
-import {InternalConfig} from './utils';
+import {WorkbenchExtension} from './WorkbenchExtension';
+
 
-interface PackageInfo {
+interface ExtensionInfo {
   name: string;
   version: string;
   aiKey: string;
 }
 
-
-export interface TelemetryContext {
-  properties: TelemetryProperties;
-  measurements: TelemetryMeasurements;
+/**
+ * Operation result of telemetry
+ */
+export enum TelemetryResult {
+  Succeeded = 'Succeeded',
+  Failed = 'Failed',
+  Cancelled = 'Cancelled',
 }
 
-export interface TelemetryProperties {
-  result: 'Succeeded'|'Failed'|'Cancelled';
-  error: string;
-  errorMessage: string;
-  [key: string]: string;
-}
-
-export interface TelemetryMeasurements {
-  duration: number;
-  [key: string]: number;
-}
-
-
-function getPackageInfo(context: vscode.ExtensionContext): PackageInfo|
-    undefined {
-  const extensionPackage = require(context.asAbsolutePath('./package.json'));
-  if (extensionPackage) {
-    const packageInfo: PackageInfo = {
-      name: extensionPackage.name,
-      version: extensionPackage.version,
-      aiKey: extensionPackage.aiKey,
-    };
-    return packageInfo;
-  }
-  return undefined;
+/**
+ * Context of telemetry
+ */
+export interface TelemetryContext {
+  properties: {[key: string]: string};
+  measurements: {[key: string]: number};
 }
 
 export class TelemetryWorker {
-  private static _reporter: TelemetryReporter;
-
-  static sendEvent(eventName: string, telemetryContext: TelemetryContext):
-      void {
-    if (this._reporter) {
-      this._reporter.sendTelemetryEvent(
-          eventName, telemetryContext.properties,
-          telemetryContext.measurements);
-    }
-  }
-
-  static async dispose() {
-    if (this._reporter) {
-      await this._reporter.dispose();
-    }
-  }
+  private _reporter: TelemetryReporter|undefined;
+  private extensionContext: vscode.ExtensionContext|undefined;
 
-  static initialize(context: vscode.ExtensionContext): void {
-    const packageInfo = getPackageInfo(context);
-    if (!packageInfo) {
+  constructor(context: vscode.ExtensionContext) {

```

Comment:
Why do we want to remove singleton from TelemetryWorker? I may be lack of context, let's discuss next week.

Reply:
Fixed to be singleton.